lib/Basic/Unix/TaskQueue.inc

```diff
@@ -249,45 +247,48 @@ bool Task::execute() {
   return false;
 }
 
-static bool readFromAPipe(int Pipe, std::string &Output,
-                          UnifiedStatsReporter *Stats,
-                          bool UntilEnd) {
+/// \p Pipe must be in blocking mode, and must contain unread data.
+/// Read the data in \p Pipe, and append it to \p Output.
+/// If \p UntilEnd is true, keep reading, and possibly hanging, till the pipe is
+/// closed. If \p UntilEnd is false, just read once. \return true if error
+static bool readFromAPipe(std::string &Output, int Pipe,
+                          UnifiedStatsReporter *Stats, bool UntilEnd) {
   char outputBuffer[1024];
   ssize_t readBytes = 0;
   while ((readBytes = read(Pipe, outputBuffer, sizeof(outputBuffer))) != 0) {
     if (readBytes < 0) {
       if (errno == EINTR)
         // read() was interrupted, so try again.
+        // FIXME: Should there be a counter to break out of this loop if
+        // there are more than some number of EINTRs?
         continue;
       return true;
     }
-
     Output.append(outputBuffer, readBytes);
     if (Stats)
       Stats->getDriverCounters().NumDriverPipeReads++;
     if (!UntilEnd)
       break;
   }
-
   return false;
 }
 
 bool Task::readFromPipes(bool UntilEnd) {
-  bool Ret = readFromAPipe(Pipe, Output, Stats, UntilEnd);
+  bool Ret = readFromAPipe(Output, Pipe, Stats, UntilEnd);
   if (SeparateErrors) {
-    Ret |= readFromAPipe(ErrorPipe, Errors, Stats, UntilEnd);
+    Ret |= readFromAPipe(Errors, ErrorPipe, Stats, UntilEnd);
   }
   return Ret;
 }
 
 void Task::finishExecution() {
-  assert(State == Executing &&
+  assert(State == TaskState::Executing &&
          "This Task must be executing to finish execution!");
 
-  State = Finished;
+  State = TaskState::Finished;
 
   // Read the output of the command, so we can use it later.
-  readFromPipes();
+  readFromPipes(/*UntilEnd = */ false);
```

Comment:
Nitpick: we usually format this as `/*UntilEnd*/false`.

Reply:
Fixed.